### PR TITLE
Merge master into branch-2.0

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Merge master into branch-2.0.

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/R/mlflow/R/install.R [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# Returns the current MLflow R package version
```
Comment (harupy) [inline mlflow/R/mlflow/R/install.R:1]: TODO: Remove this file.

**Comment by harupy** [comment]
@mlflow-automation autoformat

File: mlflow/__init__.py [hunk 2]
```diff
@@ -127,48 +124,53 @@
...
+    active_run,
+    get_run,
+    start_run,
+    end_run,
+    search_runs,
+    list_run_infos,
+    get_artifact_uri,
+    get_experiment,
+    get_experiment_by_name,
+    list_experiments,
```
Comment (harupy) [inline mlflow/__init__.py:148]:
```suggestion
```

File: mlflow/__init__.py [hunk 3]
```diff
@@ -127,48 +124,53 @@
...
+    log_text,
+    log_dict,
+    log_image,
+    log_figure,
+    active_run,
+    get_run,
+    start_run,
+    end_run,
+    search_runs,
+    list_run_infos,
```
Comment (harupy) [inline mlflow/__init__.py:144]:
```suggestion
```

File: mlflow/models/__init__.py [hunk 4]
```diff
@@ -31,49 +31,6 @@
...
-    """
-    get_flavor_backend(model_uri, docker_build=True, env_manager=env_manager).build_image(
-        model_uri,
-        name,
-        mlflow_home=mlflow_home,
-        install_mlflow=install_mlflow,
-        enable_mlserver=enable_mlserver,
-    )
-
-
```
Comment (harupy) [inline mlflow/models/__init__.py:76]: This needs to be reverted

File: mlflow/__init__.py [hunk 5]
```diff
@@ -57,9 +53,11 @@
     from mlflow import fastai
     from mlflow import gluon
     from mlflow import h2o
+    from mlflow import keras
```
Comment (harupy) [inline mlflow/__init__.py:56]:
```suggestion
```

File: mlflow/__init__.py [hunk 6]
```diff
@@ -65,6 +65,7 @@
     from mlflow import spark
     from mlflow import statsmodels
     from mlflow import tensorflow
+    from mlflow import keras
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/__init__.py]: Yes, I removed it.

File: mlflow/store/db/utils.py [hunk 7]
```diff
@@ -153,6 +153,10 @@ def _upgrade_db(engine):
         command.upgrade(config, "heads")
 
 
+def _upgrade_url(url):
+    _upgrade_db(create_sqlalchemy_engine_with_retry(url))
```
Comment (WeichenXu123) [inline mlflow/store/db/utils.py]: In next commits the function was removed. But how does it address incompatible db issue if without upgrading db ?

Comment (WeichenXu123) [reply mlflow/store/db/utils.py]: Shall we use file store in these mlflow 1.28 tests instead ?

Comment (harupy) [reply mlflow/store/db/utils.py]: @WeichenXu123 You're commenting to an old commit. I already revert this change.

Comment (harupy) [reply mlflow/store/db/utils.py]: > But how does it address incompatible db issue if without upgrading db ?

We can use a file store.

File: mlflow/pyfunc/__init__.py [hunk 8]
```diff
@@ -207,44 +207,55 @@
...
-    _add_code_from_conf_to_system_path,
-    _get_flavor_configuration_from_ml_model_file,
-    _validate_and_prepare_target_save_path,
+from mlflow.utils import (
+    PYTHON_VERSION,
+    get_major_minor_py_version,
+    _is_in_ipython_notebook,
 )
+from mlflow.utils import env_manager as _EnvManager
+from mlflow.utils import find_free_port
```
Comment (BenWilson2) [inline mlflow/pyfunc/__init__.py:255]: nit: should be included in bulk import line 249

**Review by BenWilson2 — APPROVED** [review]
LGTM

File: dev/Dockerfile.test [hunk 9]
```diff
@@ -10,12 +10,18 @@ RUN apt-get update && \
...
+ENV PATH=/opt/conda/bin:$PATH
+
 # Install required python packages first
 ADD requirements /app/requirements
 RUN pip install -r requirements/test-requirements.txt -r requirements/lint-requirements.txt --no-cache-dir
 
 # Add the rest of files and install mlflow in editable form + packages necessary for development
 ADD . /app
-RUN pip install --no-cache-dir -e .
+RUN pip install --no-cache-dir -e .[pipelines]
```
Comment (dbczumar) [inline dev/Dockerfile.test:25]: We removed the `[pipelines]` extra in `branch-2.0`, right?

Comment (harupy) [reply dev/Dockerfile.test:25]: Thanks for the catch! Reverted this change.

File: mlflow/ml-package-versions.yml [hunk 10]
```diff
@@ -303,17 +303,18 @@ spark:
...
+
+      # # TODO: Fix spark datasource autologging and test it against non-preview versions of pyspark
+      # if [ "$(pip show pyspark | grep Version | cut -d' ' -f2)" = "3.0.0" ]; then
+      #   YELLOW='\033[33;1;4m'
+      #   NO_COLOR='\033[0m'
+      #   echo -e "${YELLOW}WARNING: Spark datasource autologging currently only works for Spark 3.0.0-preview.${NO_COLOR}"
+      #   pip uninstall -y pyspark
+      #   ./dev/setup-spark-datasource-autologging.sh
+      #   find tests/spark/autologging/datasource -name 'test*.py' | xargs -L 1 pytest
+      # fi
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:317]: For debugging. Will revert.

**Comment by harupy** [comment]
Just running `pytest tests/spark/autologging/datasource/test_spark_datasource_autologging.py` worked fine.

**Comment by harupy** [comment]
Testing datasource autologging locally in a docker container.

**Comment by harupy** [comment]
`Cross version tests / test (spark / autologging / 3.0.0) (pull_request)` succeeded. I did nothing.

**Comment by harupy** [comment]
@dbczumar @WeichenXu123 Could you take another look?

**Review by WeichenXu123 — APPROVED** [review]

**Pull request merged**